### ISSUE TITLE
Union of imported STLs fail to render with CGAL

### ISSUE THREAD

**Issue by maradatscha** [openscad/openscad#802]
The file [crashing.scad](https://gist.github.com/maradatscha/0162ee58738ce9dfebe2) crashes OpenSCAD version 2014.05.17 when rendering (F6).
The [CrashLog](https://gist.github.com/maradatscha/560ee22d6c9f9a3692cb)

The code leading to the crash seems to be

```
CGALUtils::applyOperator(std::list<std::pair<AbstractNode const*, boost::shared_ptr<Geometry const> >, std::allocator<std::pair<AbstractNode const*, boost::shared_ptr<Geometry const> > > > const&, CGAL_Nef_polyhedron&, OpenSCADOperator) + 700
GeometryEvaluator::applyToChildren3D(AbstractNode const&, OpenSCADOperator) + 330
GeometryEvaluator::applyToChildren(AbstractNode const&, OpenSCADOperator) + 240
GeometryEvaluator::visit(State&, CsgNode const&) + 309
```

causing a `CGAL::assertion_fail` somewhere around `CGAL::Halffacet_geometry
`. I'm trying to create a minimal example but I'm failing. 
I suspect that it has something to do with the intersection between polyhedron and polyhedron or between a cube and a polyhedron.

I'll edit the files to create a smaller example if possible.

Does somebody know this error? 

After editing the scadfile to remove some boxes I'm getting

```
Rendering Polygon Mesh using CGAL...
CGAL error in CGALUtils::applyBinaryOperator union: CGAL ERROR: assertion violation!
Expr: target(s1).vertex()==target(s2).vertex()
File: ../libraries/install/include/CGAL/Nef_3/SNC_FM_decorator.h
Line: 218
```
##

**Comment by GilesBathgate** [comment]
The minimal example provided by kintel doesn't seem to produce the same exception for me, instead I get: 

```ERROR: CGAL error in CGALUtils::applyBinaryOperator union: CGAL ERROR: assertion violation! Expr: itl != it->second.end() File: /usr/include/CGAL/Nef_3/SNC_external_structure.h Line: 1125 ```

This version of the OP's script (cut down) produces the originally reported exception message:
[issue108-cut-down.scad](https://gist.github.com/GilesBathgate/065eed43ee1485af1075)

**Comment by DrLex0** [comment]
I also get the same assertion violation “Expr: itl != it->second.end()” when trying to combine two rather detailed polyhedra. They do not touch. First it was worse, even a single polyhedron would fail if I would try to combine it with something as simple as a cube. After spending many hours cleaning up the models both manually and with both Netfabb and MeshLab, they would work on their own but still give that assertion violation when combined. Eventually I gave up and included all variations of the models as separate polyhedra with some if/else statements. Horribly inefficient, but at least it is usable.

**Comment by MichaelAtOz** [comment]
@DrLex0 care to share the polyhedra, see if they are integral?

**Comment by DrLex0** [comment]
Sure: [wrench.scad.zip](https://github.com/openscad/openscad/files/1085248/wrench.scad.zip)

These are the parts from [Thingiverse thing 2102811](https://www.thingiverse.com/thing:2102811). I converted them to SCAD with a slightly modified version of [this script](https://github.com/joshuaflanagan/stl2scad), to allow rounding the coordinates to a certain precision. I checked whether the rounding causes the problem: it does not.
The meshes are certainly manifold according to Blender and Slic3r, and repairing them in Netfabb does not make a difference. The ones included in this scad file were also treated with the ‘Remove T-Vertices by Edge-Flip’ from MeshLab, but it doesn't help. I can combine each of the separate ‘body’ and ‘inserts’ polyhedra with objects generated from within OpenSCAD, but as soon as I try to render both these polyhedra together into a single result, it fails. If I merge the meshes and then convert them to a single polyhedron, it works fine.

**Comment by RonaldoCMP** [comment]
There is something odd with the polyhedron in module insert(). There is a
thin plate bellow the helical gear. It seems that this plate has
intersection with the gear but both are parts of the same polyhedron
insert(). A polyhedron with a self-intersection is not a manifold. The
images bellow show a zoom of that intersection and some faces have more
than three vertices although the polyhedron describes just triangular faces.

[image: Imagem inline 1]



​

2017-06-19 11:38 GMT-03:00 Alexander Thomas <notifications@github.com>:

**Comment by DrLex0** [comment]
Indeed, the top of the plate and bottom of the wheel had practically the same Z coordinates, which caused this mess, although none of the programs reported this as non-manifold. However, even with this fixed, I still get the assertion violation: [wrench2.scad.zip](https://github.com/openscad/openscad/files/1086153/wrench2.scad.zip)

**Comment by RonaldoCMP** [comment]
I have successfully rendered each module, body_only() and inserts(),
separately with a subtracting cube which would suggest they are valid
manifold (version 2017.01.20). However, the render of their union still
produces an access violation by CGAL.
Surprisingly:

difference(){
    inserts();
    body_only();
}

renders fine. It seems to be a CGAL bugs.

2017-06-19 16:53 GMT-03:00 Alexander Thomas <notifications@github.com>:

**Comment by MichaelAtOz** [comment]
Here is a clue
```
body_only();
translate([-40,0,0]) inserts();
```
Does not get the CGAL error.

I had to chop-up the spanner to look inside:
![spanner cut up](https://user-images.githubusercontent.com/1972961/27358638-ebf30ee6-565a-11e7-96cc-e9fe3f60166b.PNG)

The worm gear and support plate have degenerated triangles
![spanner worm degen](https://user-images.githubusercontent.com/1972961/27358646-f4aae8e2-565a-11e7-867e-a1dc754141ad.PNG)

As does the corner of the head
![spanner head corner degen](https://user-images.githubusercontent.com/1972961/27358649-f7a34c1a-565a-11e7-83f8-87052589c5bc.PNG)

And the inside of the cylinder
![spanner degen](https://user-images.githubusercontent.com/1972961/27358655-fadd115e-565a-11e7-8355-a019b9605971.PNG)

I've always had CGAL issues with imported STLs with degenerated triangles, so I assume internally generated (no pun intended) ones are the same.

**Comment by MichaelAtOz** [comment]
BTW anyone looking at the volume count 'inconsistency', there are hollow inside volumes in the chunky support pieces, e.g. the red cube above, left behind when I deleted the piece itself. and possibly elsewhere, I didn't chop-up every piece.

**Comment by RonaldoCMP** [comment]
@Michael

It is weird that a simple displacement of inserts() changes CGAL behavior
regarding degenerated triangles. I had checked for degenerated triangles
with:

    epsilon = 1e-5;

    function unit(v) = v/norm(v);

    for(tri=faces) {
        p0 = points[tri[0]];
        p1 = points[tri[1]];
        p2 = points[tri[2]];
        cr = cross(unit(p0-p1), unit(p0-p2));
        if(norm(cr)<epsilon) echo(tri, norm(cr));
    }

One triangle of body_only() and 3 triangles of inserts() were reported as
having a cr of the order of 1e-6. What is a degenerated triangles to CGAL?

How did you split the connected components of the model?


2017-06-20 19:33 GMT-03:00 Michael <notifications@github.com>:

> BTW anyone looking at the volume count 'inconsistency', there are hollow
> inside volumes in the chunky support pieces, e.g. the red cube above, left
> behind when I deleted the piece itself. and possibly elsewhere, I didn't
> chop-up every piece.
>

**Comment by MichaelAtOz** [comment]
I used Netfabb basic, I just checked, It regards an edge length less than 0.001mm as bad (rightly or wrongly). I tried making that smaller, but next step was to zero...

However, when I see something like the red circled bit:
![spanner anomaly cut](https://user-images.githubusercontent.com/1972961/27363927-b795ce16-567a-11e7-8d68-3a2affb4f276.PNG)
It makes me look closer. (blue bit see below)

As I said, whenever I get CGAL issues with a STL, mostly Netfabb show degeneraged faces, as in the previous post.

I suspect CGAL, not my specialty, may be able to get away with one part of a binary operation degenerate, but two maybe not??? 

> How did you split the connected components of the model?

Netfabb has a handy cut operation, the blue bit above. You move the X Y Z slider, execute cut, the next step allows you to either fill the open cut, or leave it open so you see inside (like I did with the handle above). You then select one of the cut pieces and menu Part/Move to translate it apart, so you can look between the pieces. Rinse & repeat to chop your way into the inner sanctum...Sometimes you need to select a bunch of faces and remove them to get a view. 

Sort of a combination of fine surgery and a crow bar...

**Comment by nophead** [comment]
>It is weird that a simple displacement of inserts() changes CGAL behavior regarding degenerated triangles

Floating point numbers don't have constant precision. As you move away from zero the resolution gets less so an offset can cause faces to degenerate if they are very close.

**Comment by GilesBathgate** [comment]
@nophead Does openscad not convert the stl to Gmpq numbers and then translate those?

**Comment by nophead** [comment]
I could be wrong but I think OpenSCAD only uses CGAL as a last resort. So if you import an STL and translate it I think it will be done in floats or doubles. When it needs to do a union it will convert to CGAL's number system at that point.

**Comment by GilesBathgate** [comment]
@nophead I guess its a performance vs robustness trade-off. I don't think transformations are not a bottleneck though, at least not  in my experience. Its the CGAL union that's the performance problem.

**Comment by nophead** [comment]
Yes but I don't think it knows it is going to be used in union until it is. A lot of the 3D stuff is done outside of CGAL these days using a different polyhedron representation. This is why you can have non-manifold objects that look fine but as soon as you union them with a cube you get a CGAL error.

**Comment by GilesBathgate** [comment]
@nophead Ah, that's cool. I moved away from the "point-cloud" based representation a while back, and moved to "triangles-with-pointers", (I am sure there is better terminology) Gmpq is usable outside of cgal however, which is what I use to back the polyhedron representation, and all mathematical calculations.

**Comment by nophead** [comment]
Seems like a good idea to have a consistent number representation. OpenSCAD uses at least 3.

**Comment by RonaldoCMP** [comment]
I am not convinced of it. The object inserts() is around  40 units from the
origin and Michael's translation adds 40 units more to that. How this could
make triangles lose degenerecence? I would expect the reverse. And as I
said, if we take the cross product of the unit vectors of two triangle
edges we get something of order 1e-6. Is this value low enough to consider
the triangle degenerate? The translation seems have no numerical influence
on the cross product values.

2017-06-21 5:06 GMT-03:00 Chris <notifications@github.com>:

**Comment by GilesBathgate** [comment]
Sometimes it is worth reducing the STL to the simplest number of triangles that still produce the degenerate case, I did this once I think and made some progress on #1580 IIRC

**Comment by nophead** [comment]
@RonaldoCMP,
   If you double the displacement from the origin the floating point exponent increases by one and the bottom bit of the mantissa now has twice the significance. If you had two points that only differed by the bottom bit they would now be coincident and a triangle could collapse.

If one of the object's points was originally close to the origin and is now 40 units away you could lose a lot more bits.

**Comment by RonaldoCMP** [comment]
I understand that but in the Alexander's case, no edge has a length lesser
than 1e-4.

I found, however, two identical points in the inserts definition: points[0]
and points[190]. Although this cannot be blamed as the source of a
non-manifold, it should be further investigated.

2017-06-21 17:36 GMT-03:00 Chris <notifications@github.com>:

> @RonaldoCMP <https://github.com/ronaldocmp>,
> If you double the displacement from the origin the floating point exponent
> increases by one and the bottom bit of the mantissa now has twice the
> significance. If you had two points that only differed by the bottom bit
> they would now be coincident and a triangle could collapse.
>
> If one of the object's points was originally close to the origin and is
> now 40 units away you could lose a lot more bits.
>

**Comment by MichaelAtOz** [comment]
If you translate body_only() +40 instead of inserts() -40 there are still no CGAL errors.

**Comment by jbcpollak** [comment]
I'm having this problem with 2019.1-RC4:

```
CGAL error: assertion violation!
Expression : itl != it->second.end()
File       : /Users/kintel/code/OpenSCAD/libraries/install/include/CGAL/Nef_3/SNC_external_structure.h
Line       : 1153
Explanation: 
Refer to the bug-reporting instructions at https://www.cgal.org/bug_report.html
ERROR: CGAL error in CGALUtils::applyBinaryOperator difference: CGAL ERROR: assertion violation!
Expr: itl != it->second.end()
File: /Users/kintel/code/OpenSCAD/libraries/install/include/CGAL/Nef_3/SNC_external_structure.h
Line: 1153
CGAL error: assertion violation!
Expression : itl != it->second.end()
File       : /Users/kintel/code/OpenSCAD/libraries/install/include/CGAL/Nef_3/SNC_external_structure.h
Line       : 1153
Explanation: 
Refer to the bug-reporting instructions at https://www.cgal.org/bug_report.html
ERROR: CGAL error in CGALUtils::applyBinaryOperator union: CGAL ERROR: assertion violation!
Expr: itl != it->second.end()
File: /Users/kintel/code/OpenSCAD/libraries/install/include/CGAL/Nef_3/SNC_external_structure.h
Line: 1153
Current top level object is empty.
```

I am importing an STL, they attempting to do a difference operation on it. The STL was generated by OpenSCAD immediately before being re-imported by another .scad file.

As a work around, is there any other way to memoize or cache rendered output from OpenSCAD so that future operations will run faster? I want to generate a base object once and then apply ~100 differences to the object. The difference operation doesn't take very long, but generating the object that many times would be prohibitively slow. So what I ended up doing is saving the object to STL, and then writing scripts to re-import it. In my usecase the 'base object' is actually potential one of 100s, so being able to cache a large set of them in a rendered form would be very useful.

**Comment by MichaelAtOz** [comment]
First, try Design/Flush-cache before doing F6 (without F5) then export.
There is some issue with F5 prior to F6 in some cases.

Alternatively use `render()` to wrap your base objects, the CGAL geometry is then cached, until you exit or flush.

**Comment by t-paul** [comment]
Using a more structured file format might help too, e.g. 3MF (should be enabled in the release candidates for upcoming release).

**Comment by jbcpollak** [comment]
> This seems to have fixed/worked around the problem. Thanks!

Spoke too soon, with a different base 3MF imported, I get this error:

```
\CGAL error: assertion violation!
Expression : target(s1).vertex()==target(s2).vertex()
File       : /Users/kintel/code/OpenSCAD/libraries/install/include/CGAL/Nef_3/SNC_FM_decorator.h
Line       : 228
Explanation: 
Refer to the bug-reporting instructions at https://www.cgal.org/bug_report.html
ERROR: CGAL error in CGALUtils::applyBinaryOperator difference: CGAL ERROR: assertion violation!
Expr: target(s1).vertex()==target(s2).vertex()
File: /Users/kintel/code/OpenSCAD/libraries/install/include/CGAL/Nef_3/SNC_FM_decorator.h
Line: 228
```

I see there are a ton of open inter-related tickets around imported STLs and CGAL.

**Comment by jbcpollak** [comment]
I made minimal examples of these problems and uploaded them here: https://github.com/jbcpollak/openscad-problems

See the `3mf` and `stl` directories to see the problems, just run the shell scripts in each directory to demonstrate the problems.

**Comment by DTJF** [comment]
I can provide a simple .scad file causing that issue (version 2019.05):

a custom nut compiles and previews fine, but doesn't render
![grafik](https://user-images.githubusercontent.com/11519207/72593655-c7ec6700-3905-11ea-81c9-3d88c8b460c5.png)
### Source code:
~~~~
module IGewinde(r1, r2, n, h)
{
  r=(r1+r2*r2/r1)/2.141;
  linear_extrude(height=h, twist=-360*h/n)
  difference(){
    circle(r2);
    translate([0,r-r1,0]) circle(r);
  }
}

module mutter(r1=10.4, r2=11.8, r3=14.5, r4=9.5, r5=14
            , s=12, n=1.5, h1=1, h2=9, h3=2)
{
  h=h1+h2;
  union(){
    difference(){
      cylinder(r=r3, h=h2+h1, $fn=s);
      translate([0,0,-h1]) cylinder(r=r2, h=h2+3*h1);
    }
    translate([0,0,h1]) IGewinde(r1, r2, n, h2);
    rotate_extrude()
    polygon(points=[
      [r2,h],[r4,h],[r4,h+h3],[r5-1,h+h3],[r5,h+1],[r5,h]
    ]);
  }
}
mutter();
~~~~
### Error message:
Parsing design (AST generation)...
Saved backup file: ~/.local/share/OpenSCAD/backups/Mutter-backup-AezdxICS.scad
Compiling design (CSG Tree generation)...
Rendering Polygon Mesh using CGAL...
**ERROR: CGAL error in CGALUtils::applyBinaryOperator union: CGAL ERROR: assertion violation! Expr: itl != it->second.end() File: /usr/include/CGAL/Nef_3/SNC_external_structure.h Line: 1106**
Geometries in cache: 22
Geometry cache size in bytes: 5894288
CGAL Polyhedrons in cache: 7
CGAL cache size in bytes: 113616
Total rendering time: 0 hours, 0 minutes, 46 seconds
Rendering finished.

**Comment by RonaldoCMP** [comment]
The problem seems to be in IGewinde that doesn't pass CGAL requirements
when unioined to a cube in render.

IGewinde(10.4, 11.8, 1.5, 9);
cube(20); // ERROR: CGAL error

The polygon linear_extruded in the module is almost degenerated with very
thin tips. It is the difference of two almost tangent circles.
The polygon itself may be unioned to a square and rendered but its linear
extrusion fails.

Perhaps, linear_extrude should issue a warning for bad shaped polygons.


Em sex., 17 de jan. de 2020 às 07:55, Projects around the FreeBASIC
compiler <notifications@github.com> escreveu:

**Comment by DTJF** [comment]
Indeed, when I cut the tips it renders:
~~~~
module IGewinde(r1, r2, n, h)
{
  r=(r1+r2*r2/r1)/2.141;
  linear_extrude(height=h, twist=-360*h/n)
  difference(){
    circle(r2);
    translate([0,r-r1,0]) circle(r);
    polygon(points=[[0,0],[r2,r2],[-r2,r2]]);
  }
}
~~~~
Thanks for pointing to the issue! This granularity is fine enough.

I didn't find any hint that linear_extrude doesn't like thin tips. The warning seams to be a good idea, at least in the documentation.

**Comment by kintel** [comment]
Status as of 2024-03-24: All the mentioned examples now work when using Manifold (IGewinde, addText, wrench).
